src/app/app.ts

```diff
@@ -1,71 +1,141 @@
-        if (!zdToken) {
-            throw new Error('Failed to get user access_token');
-        }
+export interface IApp {
+    createTicketFromPost(): Promise<AppCallResponse>;
+    createZDSubscription(): Promise<AppCallResponse>;
+}
 
+class App implements IApp {
+    context: AppContext
+    values: AppCallValues
+
+    constructor(call: AppCall) {
+        this.context = call.context;
+        this.values = call.values;
+    }
+    createTicketFromPost = async (): Promise<AppCallResponse> => {
         // get zendesk client for user
-        const zdClient = newZDClient(zdToken);
+        const zdClient = newZDClient(this.context);
 
         // create the ticket object from the form response
-        const [zdTicket, errors] = newTicketFromForm(call.values);
+        const [zdTicketPayload, fieldErrors] = newTicketFromForm(this.values);
 
         // respond with errors
-        if (Object.keys(errors).length !== 0) {
-            return errors;
+        if (this.hasFieldErrors(fieldErrors)) {
+            return newErrorCallResponseWithFieldErrors(fieldErrors);
         }
 
         // create the ticket in Zendesk
-        const ticket = await tryPromiseWithMessage(zdClient.tickets.create(zdTicket), 'Failed to create Zendesk ticket');
+        const createReq = zdClient.tickets.create(zdTicketPayload);
+        const zdTicket = await tryPromiseWithMessage(createReq, 'Failed to create Zendesk ticket');
 
         // get the Zendesk user
-        const zdUser = await tryPromiseWithMessage(zdClient.users.show(ticket.requester_id), 'Failed to get Zendesk user');
+        const getUserReq = zdClient.users.show(zdTicket.requester_id);
+        const zdUser = await tryPromiseWithMessage(getUserReq, 'Failed to get Zendesk user');
 
         // create a reply to the original post noting the ticket was created
-        const id = ticket.id;
-        const subject = ticket.subject;
+        const id = zdTicket.id;
+        const subject = zdTicket.subject;
         const message = `${zdUser.name} created ticket [#${id}](${Env.ZD.Host}/agent/tickets/${id}) [${subject}]`;
-        await this.createBotPost(call.context, message);
+        await this.createBotPost(message);
 
-        // respond with no errors
-        return {};
+        return newOKCallResponse();
     }
 
-    createBotPost = async (context: AppContext, message: string): Promise<void> => {
-        const adminToken = configStore.getAdminAccessToken();
-        const adminClient = newMMClient(adminToken);
+    createZDSubscription = async (): Promise<AppCallResponse> => {
+        // get zendesk client for user
+        const zdClient = newZDClient(this.context);
+
+        // create the trigger object from the form response
+        let zdTriggerPayload: any;
+        try {
+            zdTriggerPayload = newTriggerFromForm(this.context, this.values);
+        } catch (e) {
+            return newErrorCallResponseWithMessage(e.message);
+        }
+
+        let request: any;
```

Comment:
Can we type this?

Reply:
put into catch-all NITs ticket https://mattermost.atlassian.net/browse/MM-33787